lib/tests/test_signature.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright 2021 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use jujutsu_lib::backend::{MillisSinceEpoch, Signature, Timestamp};
+
+#[test]
+fn test_username() {
+    let timestamp = Timestamp {
+        timestamp: MillisSinceEpoch(42),
+        tz_offset: 0,
+    };
+
+    assert_eq!(
+        Signature {
+            name: "name".to_owned(),
+            email: "username@example.com".to_owned(),
+            timestamp: timestamp.clone(),
+        }
+        .username(),
+        Some("username"),
+    );
+    assert_eq!(
+        Signature {
+            name: "name".to_owned(),
+            email: "user.name@example.com".to_owned(),
+            timestamp: timestamp.clone(),
+        }
+        .username(),
+        Some("user.name"),
+    );
+    assert_eq!(
+        Signature {
+            name: "name".to_owned(),
+            email: "x@example.com".to_owned(),
+            timestamp: timestamp.clone(),
+        }
+        .username(),
+        Some("x"),
+    );
+    assert_eq!(
+        Signature {
+            name: "name".to_owned(),
+            email: "user.name+tag@example.com".to_owned(),
+            timestamp: timestamp.clone(),
+        }
+        .username(),
+        Some("user.name+tag"),
+    );
+
+    assert_eq!(
+        Signature {
+            name: "name".to_owned(),
+            email: "invalid.email".to_owned(),
+            timestamp: timestamp.clone(),
+        }
+        .username(),
+        None,
+    );
+    assert_eq!(
+        Signature {
+            name: "name".to_owned(),
+            email: "a@b@example.com".to_owned(),
+            timestamp,
+        }
+        .username(),
+        None,
+    );
+}
```

Comment:
I think unit tests are typically in the source file itself in Rust, right? If you think they belong here, update the copyright year.